### PR TITLE
[stable] Fix Bugzilla 24599 - Wrongly elided TypeInfo emission

### PR DESCRIPTION
**Pull request by kinke**
Reverting #14844, which caused such missing TypeInfos, *and* making sure the special TypeInfo members are fully analyzed and ready for codegen (otherwise hitting an assertion).

---

Currently just checking if it breaks CI somewhere.

### CODE REVIEW

File: compiler/src/dmd/todt.d [hunk 1]
```diff
@@ -1437,6 +1437,13 @@ private extern (C++) class TypeInfoDtVisitor : Visitor
                 /* ti.toObjFile() won't get called. So, store these
                  * member functions into object file in here.
                  */
+
+                if (sd.semanticRun < PASS.semantic3done)
+                {
+                    import dmd.semantic3 : semanticTypeInfoMembers;
+                    semanticTypeInfoMembers(sd);
```
Comment (maxhaton) [inline compiler/src/dmd/todt.d:1444]: Is this a sensible place to run sema

**Comment by RazvanN7** [comment]
@kinke is there a bugzilla issue for this? Or do you have a test that I can reproduce? I'm kind of surprised that this doesn't break any code at there, since IIRC this was needed to advance work on templating the druntime hooks.

**Comment by kinke** [comment]
@RazvanN7: Nope sorry, no issue yet; the problem occurs in a private codebase, and unfortunately one of the largest/most complicated ones at Symmetry, so I'm not exactly looking forward to setting up an involved dustmite session. ;)

**Comment by RazvanN7** [comment]
I see. I'm just wondering what's the best way to proceed here. If I merge the revert, I'm gonna have to find a different fix for Issue 23650, but without a test case I risk on breaking your code again. If I don't merge it, your code stays broken.

**Comment by kinke** [comment]
> I'm gonna have to find a different fix for Issue 2365

I think this can only be properly fixed by re-designing the way struct TypeInfos (and the special TypeInfo member functions) are emitted in DMD, adopting the LDC way (I've changed this some years ago, no more TypeInfo issues since then AFAICT: https://github.com/ldc-developers/ldc/pull/3491). The main change there is that the struct TypeInfo emission has been moved to the codegen/glue layer - whenever the symbol is needed *during codegen* (so no CTFE usages, or usages in purely speculative contexts), it is emitted into the referencing object file (as a COMDAT, so that the linker folds duplicates across object files), instead of trying to decide whether to add a `TypeInfoDeclaration` to module members *in the frontend*, with brittle `isSpeculativeType()` (unused by LDC) logic etc.

LDC emits the special TypeInfo member functions (`xopEquals`, `xopCmp`, `xtoHash`) into the object file containing the struct declaration, so that those are guaranteed to be available. Only the `TypeInfo_Struct` symbol is lazily emitted into every referencing object file.

**Comment by kinke** [comment]
@RazvanN7: Okay, I finally found some time and motivation to dustmite our problem, filing an issue and updating this PR accordingly.

File: compiler/src/dmd/todt.d (hunk 1 above)
Comment (kinke) [reply compiler/src/dmd/todt.d:1449]: We are about to emit the functions here, so we need sema3.

**Review by RazvanN7 — APPROVED** [review]

**Comment by kinke** [comment]
[just rebased onto stable to make this part of D v2.110; I was on vacation and missed the merge window]

**Pull request merged**